### ISSUE TITLE
owncloud 6.0.2 does not run under apache 2.4/php-fpm

### ISSUE THREAD

**Issue by hp4** [owncloud/core#7719]
my archlinux distribution just updated to apache 2.4. The recommended update for php was to use php-fpm and proxy-fcgi. 

Before the update owncloud 6.0.2 worked fine with apache 2.2 and php-module. After the update it does run: images/icons und css-styleheets are not loaded.

Similar problems are reported for ubuntu/debian.
https://forum.owncloud.org/viewtopic.php?f=26&t=19754

The following apache access log for the "home" /index.php and for the "files-directory" /index.php/apps/files  shows some very strange HTTP-GET requests:

GET /index.php//index.php/apps/files/css/upload.css

with double reference to index.php and "//"-separator. 

The problem might be due to the $_SERVER-environment provided by php-fpm. But owncloud should be able to run on this typical webserver-configuration.

-----                   clicking on the opencloud icon

88.67.235.79 - - [13/Mar/2014:17:40:16 +0100] "GET /index.php HTTP/1.1" 200 13344
88.67.235.79 - - [13/Mar/2014:17:40:21 +0100] "GET /apps/files/js/file-upload.js?v=d3673708f656239c4dd25f770054095d HTTP/1.1" 304 -
88.67.235.79 - - [13/Mar/2014:17:40:21 +0100] "GET /apps/files/js/jquery.iframe-transport.js?v=d3673708f656239c4dd25f770054095d HTTP/1.1" 304 -
88.67.235.79 - - [13/Mar/2014:17:40:21 +0100] "GET /apps/files/js/jquery.fileupload.js?v=d3673708f656239c4dd25f770054095d HTTP/1.1" 304 -
88.67.235.79 - - [13/Mar/2014:17:40:21 +0100] "GET /apps/files/js/jquery-visibility.js?v=d3673708f656239c4dd25f770054095d HTTP/1.1" 304 -
88.67.235.79 - - [13/Mar/2014:17:40:21 +0100] "GET /apps/files/js/filelist.js?v=d3673708f656239c4dd25f770054095d HTTP/1.1" 304 -
88.67.235.79 - - [13/Mar/2014:17:40:21 +0100] "GET /apps/files/js/fileactions.js?v=d3673708f656239c4dd25f770054095d HTTP/1.1" 304 -
88.67.235.79 - - [13/Mar/2014:17:40:21 +0100] "GET /apps/files/js/files.js?v=d3673708f656239c4dd25f770054095d HTTP/1.1" 304 -
88.67.235.79 - - [13/Mar/2014:17:40:21 +0100] "GET /apps/files/js/keyboardshortcuts.js?v=d3673708f656239c4dd25f770054095d HTTP/1.1" 304 -
88.67.235.79 - - [13/Mar/2014:17:40:21 +0100] "GET /index.php/core/js/config.js?v=d3673708f656239c4dd25f770054095d HTTP/1.1" 200 14352
88.67.235.79 - - [13/Mar/2014:17:40:21 +0100] "GET /remote.php/core.css?v=d3673708f656239c4dd25f770054095d HTTP/1.1" 304 -
88.67.235.79 - - [13/Mar/2014:17:40:21 +0100] "GET /remote.php/core.js?v=d3673708f656239c4dd25f770054095d HTTP/1.1" 304 -
88.67.235.79 - - [13/Mar/2014:17:40:21 +0100] "GET /index.php/apps/files/css/upload.css?v=d3673708f656239c4dd25f770054095d HTTP/1.1" 200 14800
88.67.235.79 - - [13/Mar/2014:17:40:21 +0100] "GET /index.php/apps/files/css/files.css?v=d3673708f656239c4dd25f770054095d HTTP/1.1" 200 14744
88.67.235.79 - - [13/Mar/2014:17:40:41 +0100] "GET /core/img/logo-wide.svg HTTP/1.1" 304 -
88.67.235.79 - - [13/Mar/2014:17:40:41 +0100] "GET /settings/img/admin.svg HTTP/1.1" 304 -
88.67.235.79 - - [13/Mar/2014:17:40:41 +0100] "GET /settings/img/personal.svg HTTP/1.1" 304 -
88.67.235.79 - - [13/Mar/2014:17:40:41 +0100] "GET /settings/img/users.svg HTTP/1.1" 304 -
88.67.235.79 - - [13/Mar/2014:17:40:41 +0100] "GET /core/img/actions/caret.svg HTTP/1.1" 304 -
88.67.235.79 - - [13/Mar/2014:17:40:41 +0100] "GET /core/img/breadcrumb.svg HTTP/1.1" 304 -
88.67.235.79 - - [13/Mar/2014:17:40:41 +0100] "GET /settings/img/help.svg HTTP/1.1" 304 -
88.67.235.79 - - [13/Mar/2014:17:40:41 +0100] "GET /core/img/actions/logout.svg HTTP/1.1" 304 -
88.67.235.79 - - [13/Mar/2014:17:40:41 +0100] "GET /core/img/places/files.svg HTTP/1.1" 304 -
88.67.235.79 - - [13/Mar/2014:17:40:41 +0100] "GET /apps/activity/img/activity.svg HTTP/1.1" 304 -
88.67.235.79 - - [13/Mar/2014:17:40:42 +0100] "GET /apps/documents/img/documents.svg HTTP/1.1" 304 -
88.67.235.79 - - [13/Mar/2014:17:40:42 +0100] "GET /core/img/places/picture.svg HTTP/1.1" 304 -
88.67.235.79 - - [13/Mar/2014:17:40:42 +0100] "GET /apps/calendar/img/calendar.svg HTTP/1.1" 304 -
88.67.235.79 - - [13/Mar/2014:17:40:42 +0100] "GET /apps/contacts/img/contacts.svg HTTP/1.1" 304 -
88.67.235.79 - - [13/Mar/2014:17:40:42 +0100] "GET /settings/img/apps.svg HTTP/1.1" 304 -
88.67.235.79 - - [13/Mar/2014:17:40:42 +0100] "GET /core/img/actions/download.svg HTTP/1.1" 304 -
88.67.235.79 - - [13/Mar/2014:17:40:42 +0100] "GET /core/img/actions/delete.svg HTTP/1.1" 304 -
88.67.235.79 - - [13/Mar/2014:17:40:42 +0100] "GET /core/img/filetypes/text.png HTTP/1.1" 304 -
88.67.235.79 - - [13/Mar/2014:17:40:42 +0100] "GET /core/img/filetypes/folder.png HTTP/1.1" 304 -
88.67.235.79 - - [13/Mar/2014:17:40:42 +0100] "GET /core/img/places/link.svg HTTP/1.1" 304 -
88.67.235.79 - - [13/Mar/2014:17:40:42 +0100] "GET /core/img/filetypes/application-pdf.png HTTP/1.1" 304 -
88.67.235.79 - - [13/Mar/2014:17:40:41 +0100] "POST /index.php/core/ajax/translations.php HTTP/1.1" 200 14856
88.67.235.79 - - [13/Mar/2014:17:40:41 +0100] "GET /index.php/core/routes.json HTTP/1.1" 200 14296
88.67.235.79 - - [13/Mar/2014:17:40:41 +0100] "GET /remote.p/core/img/noise.png?2013-12-17T15:48:20Z HTTP/1.1" 404 1865
88.67.235.79 - - [13/Mar/2014:17:40:41 +0100] "GET /cron.php HTTP/1.1" 200 20
88.67.235.79 - - [13/Mar/2014:17:40:41 +0100] "GET /remote.p/core/img/actions/search.svg?2013-12-17T15:48:20Z HTTP/1.1" 404 1865
88.67.235.79 - - [13/Mar/2014:17:40:47 +0100] "POST /index.php/core/ajax/translations.php HTTP/1.1" 200 14856
88.67.235.79 - - [13/Mar/2014:17:41:02 +0100] "POST /index.php/core/ajax/translations.php HTTP/1.1" 200 14856
88.67.235.79 - - [13/Mar/2014:17:41:07 +0100] "GET /core/img/actions/history.svg HTTP/1.1" 304 - 
88.67.235.79 - - [13/Mar/2014:17:41:07 +0100] "GET /core/img/actions/share.svg HTTP/1.1" 304 - 
88.67.235.79 - - [13/Mar/2014:17:41:07 +0100] "GET /core/img/actions/rename.svg HTTP/1.1" 304 -
88.67.235.79 - - [13/Mar/2014:17:41:02 +0100] "GET /undefined/templates/slideshow.html HTTP/1.1" 404 1865
88.67.235.79 - - [13/Mar/2014:17:41:07 +0100] "GET /index.php/core/ajax/share.php?fetch=getItemsSharedStatuses&itemType=file HTTP/1.1" 200 14464
88.67.235.79 - - [13/Mar/2014:17:41:07 +0100] "GET /remote.p/core/img/actions/delete.svg?2013-12-17T15:48:20Z HTTP/1.1" 404 1865
88.67.235.79 - - [13/Mar/2014:17:41:07 +0100] "GET /index.php/apps/files/ajax/scan.php?force=false&dir=&requesttoken=a12709e97e645b11f592 HTTP/1.1" 200 14744
88.67.235.79 - - [13/Mar/2014:17:41:10 +0100] "POST /index.php/core/ajax/translations.php HTTP/1.1" 200 14856

-----                   clicking on the files icon 

88.67.235.79 - - [13/Mar/2014:17:42:21 +0100] "GET /index.php/apps/files HTTP/1.1" 200 13960
88.67.235.79 - - [13/Mar/2014:17:42:25 +0100] "GET /index.php//index.php/apps/files/css/upload.css?v=d3673708f656239c4dd25f770054095d HTTP/1.1" 200 15416
88.67.235.79 - - [13/Mar/2014:17:42:26 +0100] "GET /index.php//index.php/core/js/config.js?v=d3673708f656239c4dd25f770054095d HTTP/1.1" 200 14968
88.67.235.79 - - [13/Mar/2014:17:42:26 +0100] "GET /index.php//remote.php/core.js?v=d3673708f656239c4dd25f770054095d HTTP/1.1" 200 14464
88.67.235.79 - - [13/Mar/2014:17:42:26 +0100] "GET /index.php//index.php/apps/files/css/files.css?v=d3673708f656239c4dd25f770054095d HTTP/1.1" 200 15360
88.67.235.79 - - [13/Mar/2014:17:42:27 +0100] "GET /index.php//remote.php/core.css?v=d3673708f656239c4dd25f770054095d HTTP/1.1" 200 14520
88.67.235.79 - - [13/Mar/2014:17:42:27 +0100] "GET /index.php//apps/files/js/file-upload.js?v=d3673708f656239c4dd25f770054095d HTTP/1.1" 200 15024
88.67.235.79 - - [13/Mar/2014:17:42:31 +0100] "GET /index.php//apps/files/js/jquery.iframe-transport.js?v=d3673708f656239c4dd25f770054095d HTTP/1.1" 200 15696
88.67.235.79 - - [13/Mar/2014:17:42:35 +0100] "GET /index.php//apps/files/js/jquery.fileupload.js?v=d3673708f656239c4dd25f770054095d HTTP/1.1" 200 15360
88.67.235.79 - - [13/Mar/2014:17:42:40 +0100] "GET /index.php//apps/files/js/jquery-visibility.js?v=d3673708f656239c4dd25f770054095d HTTP/1.1" 200 15360
88.67.235.79 - - [13/Mar/2014:17:42:45 +0100] "GET /index.php//apps/files/js/filelist.js?v=d3673708f656239c4dd25f770054095d HTTP/1.1" 200 14856
88.67.235.79 - - [13/Mar/2014:17:42:50 +0100] "GET /index.php//apps/files/js/fileactions.js?v=d3673708f656239c4dd25f770054095d HTTP/1.1" 200 15024
88.67.235.79 - - [13/Mar/2014:17:42:54 +0100] "GET /index.php//apps/files/js/files.js?v=d3673708f656239c4dd25f770054095d HTTP/1.1" 200 14688
88.67.235.79 - - [13/Mar/2014:17:42:59 +0100] "GET /index.php//apps/files/js/keyboardshortcuts.js?v=d3673708f656239c4dd25f770054095d HTTP/1.1" 200 15360
88.67.235.79 - - [13/Mar/2014:17:43:27 +0100] "GET /index.php//core/img/actions/caret.svg HTTP/1.1" 200 14912
88.67.235.79 - - [13/Mar/2014:17:43:27 +0100] "GET /index.php//settings/img/users.svg HTTP/1.1" 200 14688
88.67.235.79 - - [13/Mar/2014:17:43:27 +0100] "GET /index.php//settings/img/help.svg HTTP/1.1" 200 14632
88.67.235.79 - - [13/Mar/2014:17:43:27 +0100] "GET /index.php//core/img/logo-wide.svg HTTP/1.1" 200 14688

**Comment by wpoely86** [comment]
Sorry to bump this but with the release of owncloud 8, I would like to know if the patch in https://github.com/owncloud/core/issues/7719#issuecomment-52457150 from @niobos will ever be merge in the master? I'm running CentOS 7 and after every update from owncloud, I have to apply this patch manually to get owncloud working with php-fpm. I know it's fixed upstream in php but the fix is not backported to RHEL 7.

**Comment by jnweiger** [comment]
this patch goes into server core. Not a packaging issue. Removing the label. @DeepDiver1975 please consider.

**Comment by danmac-uk** [comment]
+1 for seeing that patch in the Ubuntu 14.04 packages ASAP

I'll probably end up doing what I did last time though. (Install ownCloud, grapple with terrible performance which pretty much breaks the experience, try to get PHP-FPM running, fail miserably, disable apache and install nginx)

One other thing to note, with PHP-FPM and that patch, my WebDAV failed to authenticate. But may just be the fact that it's 2am and I threw it into version 8 slapdash.

Cheers guys.

**Comment by wpoely86** [comment]
@danmac-uk I've got owncloud 8 running on CentOS 7 with Apache 2.4 and php-fpm. Performance is no issue. In https://github.com/owncloud/core/issues/11779#issuecomment-61063630 you can read what I had to do to get it up and running.

**Comment by danmac-uk** [comment]
@wpoely86 saved the day mate, very much appreciate you letting me know about that, thank you. It's a funny little bug this, really. None of it is ownCloud's fault (well apart from using PHP in the first place *zing* haha) but regardless, it does reflect badly on the software. Performance with the php5 Apache module is quite frankly broken - mpm_event and php5-fpm is a massive improvement. The patch above from @niobos and some documentation will surely help.

**Comment by meichenberger** [comment]
With ownCloud 8.0.2 you only need to add the following line to your vhost.conf file:
```sh
SetEnvIf Authorization "(.*)" HTTP_AUTHORIZATION=$1
```

Software versions used:
- Linux Gentoo 64-bit with 3.10.56 Kernel)
- ownCloud 8.0.2
- Apache 2.4.12
- php-fpm 5.6.6

A complete working example:
```sh
<VirtualHost *:443>
        ServerName www.example.com
        DocumentRoot "/var/www/example/com/htdocs"
        SSLEngine on

        SSLCertificateFile /etc/ssl/apache2/www.example.com.cert.pem
        SSLCertificateKeyFile /etc/ssl/apache2/www.example.com.key.pem
        SSLCertificateChainFile /etc/ssl/apache2/RapidSSL-intermediate-SHA256-CA.bundle.pem

        SetEnvIf Authorization "(.*)" HTTP_AUTHORIZATION=$1
        ProxyPassMatch ^/.*.php(/.*)?$ "unix:/var/run/php-fpm/3730083_owncloud.socket|fcgi://localhost/var/www/example/com/htdocs"
        DirectoryIndex index.php index.html

        <Directory "/var/www/example/com/htdocs">
                SSLRequireSSL
                Options FollowSymLinks IncludesNOEXEC SymlinksIfOwnerMatch
                AllowOverride All
                Require all granted
        </Directory>

        <Location />
                SSLRequireSSL
        </Location>
</VirtualHost>
```

Please update the documentation:
https://doc.owncloud.org/server/8.0/admin_manual/installation/source_installation.html

**Comment by wpoely86** [comment]
@meichenberger That's because you run Apache 2.4.12 and php 5.6.6. On RHEL 7.x (and all it's derivatives), we don't have that and that's why we need a patch. The patch of @niobos works around a bug in apache/php that's already fixed upstream but not backported to RHEL.

**Comment by meichenberger** [comment]
@wpoely86: Sorry, my bad! The update of the documentation for the more modern Linux distributions would still be nice ;-)

**Comment by RealRancor** [comment]
@meichenberger 
Requests for updating the documentation should be handled in https://github.com/owncloud/documentation/issues

**Comment by wpoely86** [comment]
@meichenberger on a slight related note, if you are running such a recent version of apache, why not use something like
```apache
<FilesMatch \.php$>
    SetHandler  "unix:/var/run/php-fpm/3730083_owncloud.socket|fcgi://localhost/var/www/example/com/htdocs"
</FilesMatch>
```
Much nicer than the `ProxyPassMatch` stuff IMHO.

**Comment by warrenseine** [comment]
Upgrading PHP to 5.6.11 worked for me.

**Comment by RealRancor** [comment]
Could anyone test https://github.com/owncloud/core/pull/18194 ?

**Comment by yardenac** [comment]
Our server still has:

Ubuntu 14.04
Apache 2.4.7-1ubuntu4.5
PHP 5.5.9-1ubuntu4.11

I don't know if the upstream fixes have been backported. Not sure how to check. Assuming they haven't.

We've been using the patch from @niobos, and it has worked so far: https://github.com/owncloud/core/issues/7719#issuecomment-40290552

But now with the update to 8.1, this patch will no longer work. The file /lib/private/request.php no longer exists. This is preventing us from upgrading.

**Comment by wpoely86** [comment]
@yardenac you can use this:
```php
diff -ur owncloud.orig/lib/private/appframework/http/request.php owncloud/lib/private/appframework/http/request.php
--- owncloud.orig/lib/private/appframework/http/request.php	2015-07-03 00:31:39.000000000 +0200
+++ owncloud/lib/private/appframework/http/request.php	2015-07-08 15:40:42.438090223 +0200
@@ -522,8 +522,18 @@
 	 */
 	public function getRequestUri() {
 		$uri = isset($this->server['REQUEST_URI']) ? $this->server['REQUEST_URI'] : '';
+
+		$script_name = $_SERVER['SCRIPT_NAME'];
+		if (array_key_exists('PATH_INFO', $_SERVER) === true) {
+			$pos = strpos($_SERVER['SCRIPT_NAME'], rawurldecode($_SERVER['PATH_INFO']));
+			if ($pos === false) {
+			} else {
+				$script_name = substr($_SERVER['SCRIPT_NAME'], 0, $pos);
+			}
+		}
+
 		if($this->config->getSystemValue('overwritewebroot') !== '' && $this->isOverwriteCondition()) {
-			$uri = $this->getScriptName() . substr($uri, strlen($this->server['SCRIPT_NAME']));
+			$uri = $this->getScriptName() . substr($uri, strlen($script_name));
 		}
 		return $uri;
 	}
@@ -548,6 +558,15 @@
 		}
 
 		$scriptName = $this->server['SCRIPT_NAME'];
+
+		if (array_key_exists('PATH_INFO', $_SERVER) === true) {
+			$pos = strpos($_SERVER['SCRIPT_NAME'], rawurldecode($_SERVER['PATH_INFO']));
+			if ($pos === false) {
+			} else {
+				$scriptName = substr($_SERVER['SCRIPT_NAME'], 0, $pos);
+			}
+		}
+
 		$pathInfo = $requestUri;
 
 		// strip off the script name's dir and file name
@@ -604,6 +623,14 @@
 	 */
 	public function getScriptName() {
 		$name = $this->server['SCRIPT_NAME'];
+		if (array_key_exists('PATH_INFO', $_SERVER) === true) {
+			$pos = strpos($_SERVER['SCRIPT_NAME'], rawurldecode($_SERVER['PATH_INFO']));
+			if ($pos === false) {
+			} else {
+				$name = substr($_SERVER['SCRIPT_NAME'], 0, $pos);
+			}
+		}
+
 		$overwriteWebRoot =  $this->config->getSystemValue('overwritewebroot');
 		if ($overwriteWebRoot !== '' && $this->isOverwriteCondition()) {
 			// FIXME: This code is untestable due to __DIR__, also that hardcoded path is really dangerous
```

**Comment by PVince81** [comment]
Is this still an issue with OC 8.0.9, 8.1.4 or 8.2.1 ?

**Comment by wpoely86** [comment]
@PVince81 I'm running 8.2.1 with the patch from above. So yes.

**Comment by RealRancor** [comment]
Might also be related to this buggy behavior of Apache and fpm: https://github.com/owncloud/core/issues/20856

**Comment by wpoely86** [comment]
@RealRancor no, what we have here is this bug: https://bugs.php.net/bug.php?id=65641

As long as RHEL & derivatives don't patch it, we're stuck using this patch.

**Comment by RealRancor** [comment]
@wpoely86 Ah, yes. Mixed those two Apache and PHP bugs. The linked post in #20856 (https://www.drupal.org/node/2603696) is mentioning that SCRIPT_NAME is also wrongly set with Apache 2.4.17

**Comment by matpen** [comment]
I am running owncloud 8.2.2 on Ubuntu 14.04 with Apache 2.4.7 and PHP-FPM 5.5.9 and I found that the patch is needed. For this version of owncloud, the patch has to be modified as follows:
```diff
--- /var/www/owncloud/lib/private/appframework/http/request.php.orig	2015-12-21 13:46:53.000000000 +0100
+++ /var/www/owncloud/lib/private/appframework/http/request.php	2015-12-30 13:47:55.136513303 +0100
@@ -576,8 +576,18 @@
 	 */
 	public function getRequestUri() {
 		$uri = isset($this->server['REQUEST_URI']) ? $this->server['REQUEST_URI'] : '';
+
+		$script_name = $_SERVER['SCRIPT_NAME'];
+		if (array_key_exists('PATH_INFO', $_SERVER) === true) {
+			$pos = strpos($_SERVER['SCRIPT_NAME'], rawurldecode($_SERVER['PATH_INFO']));
+			if ($pos === false) {
+			} else {
+				$script_name = substr($_SERVER['SCRIPT_NAME'], 0, $pos);
+			}
+		}
+
 		if($this->config->getSystemValue('overwritewebroot') !== '' && $this->isOverwriteCondition()) {
-			$uri = $this->getScriptName() . substr($uri, strlen($this->server['SCRIPT_NAME']));
+			$uri = $this->getScriptName() . substr($uri, strlen($script_name));
 		}
 		return $uri;
 	}
@@ -602,6 +612,15 @@
 		}
 
 		$scriptName = $this->server['SCRIPT_NAME'];
+
+		if (array_key_exists('PATH_INFO', $_SERVER) === true) {
+		$pos = strpos($_SERVER['SCRIPT_NAME'], rawurldecode($_SERVER['PATH_INFO']));
+			if ($pos === false) {
+			} else {
+				$scriptName = substr($_SERVER['SCRIPT_NAME'], 0, $pos);
+			}
+		}
+
 		$pathInfo = $requestUri;
 
 		// strip off the script name's dir and file name
@@ -658,6 +677,14 @@
 	 */
 	public function getScriptName() {
 		$name = $this->server['SCRIPT_NAME'];
+		if (array_key_exists('PATH_INFO', $_SERVER) === true) {
+			$pos = strpos($_SERVER['SCRIPT_NAME'], rawurldecode($_SERVER['PATH_INFO']));
+			if ($pos === false) {
+			} else {
+				$name = substr($_SERVER['SCRIPT_NAME'], 0, $pos);
+			}
+		}
+
 		$overwriteWebRoot =  $this->config->getSystemValue('overwritewebroot');
 		if ($overwriteWebRoot !== '' && $this->isOverwriteCondition()) {
 			// FIXME: This code is untestable due to __DIR__, also that hardcoded path is really dangerous

```
According to my tests, the patch is not needed with owncloud 8.2.2 on Ubuntu 15.10 with Apache 2.4.12 and PHP-FPM 5.6.11.

**Comment by yardenac** [comment]
Does this patch need to be updated for 9.0.0?

**Comment by wpoely86** [comment]
@yardenac No, it works without changes. However, other issues pop up for when upgrading to owncloud 9 (unrelated to this I think).

**Comment by Caffe1neAdd1ct** [comment]
Due to using ProxyPassMatch and PHP-FPM i'm not actually seeing a PATH_INFO variable set, quick hacky solution for my Raspberry Pi 2 Arch Linux installation with the following configs and `owncloud 9.1.0-1` installed through pacman:

/etc/httpd/conf/extra/owncloud.conf
```
<VirtualHost *:80>

    ServerName owncloud.castle
    ServerAdmin support@castle
    DocumentRoot /usr/share/webapps/owncloud/

    SetEnvIf Authorization "(.*)" HTTP_AUTHORIZATION=$1
    SetEnv HOME /usr/share/webapps/owncloud
    SetEnv HTTP_HOME /usr/share/webapps/owncloud
    
    DirectoryIndex index.php
    ProxyPassMatch ^/(.*\.php(/.*)?)$ fcgi://127.0.0.1:9001/usr/share/webapps/owncloud/$1

    #SetEnvIfNoCase ^Authorization$ "(.+)" XAUTHORIZATION=$1
    #RequestHeader set XAuthorization %{XAUTHORIZATION}e env=XAUTHORIZATION

    <Directory "/usr/share/webapps/owncloud">
        AllowOverride All
        Require all granted
    </Directory>

    ErrorLog "/var/log/httpd/owncloud.error.log"
    CustomLog "/var/log/httpd/owncloud.access.log" combined

</VirtualHost>
```

/etc/php/php-fpm.d/owncloud.conf
```
[owncloud]
user = http
group = http
listen = 127.0.0.1:9001
listen.owner = http
listen.group = http
listen.allowed_clients = 127.0.0.1
pm = dynamic
pm.max_children = 5
pm.start_servers = 2
pm.min_spare_servers = 1
pm.max_spare_servers = 3
```

Patch file for /usr/share/webapps/owncloud/lib/private/AppFramework/Http/Request.php
```
diff --git a/lib/private/AppFramework/Http/Request.php b/lib/private/AppFramework/Http/Request.php
index 1508158..994ef60 100644
--- a/lib/private/AppFramework/Http/Request.php
+++ b/lib/private/AppFramework/Http/Request.php
@@ -594,7 +594,7 @@
 	public function getRequestUri() {
 		$uri = isset($this->server['REQUEST_URI']) ? $this->server['REQUEST_URI'] : '';
 		if($this->config->getSystemValue('overwritewebroot') !== '' && $this->isOverwriteCondition()) {
-			$uri = $this->getScriptName() . substr($uri, strlen($this->server['SCRIPT_NAME']));
+			$uri = $this->getScriptName() . substr($uri, strlen($this->fixPhpFpmScriptName()));
 		}
 		return $uri;
 	}
@@ -618,7 +618,7 @@
 			$requestUri = substr($requestUri, 0, $pos);
 		}
 
-		$scriptName = $this->server['SCRIPT_NAME'];
+		$scriptName = $this->fixPhpFpmScriptName();
 		$pathInfo = $requestUri;
 
 		// strip off the script name's dir and file name
@@ -674,7 +674,7 @@
 	 * @return string the script name
 	 */
 	public function getScriptName() {
-		$name = $this->server['SCRIPT_NAME'];
+		$name = $this->fixPhpFpmScriptName();
 		$overwriteWebRoot =  $this->config->getSystemValue('overwritewebroot');
 		if ($overwriteWebRoot !== '' && $this->isOverwriteCondition()) {
 			// FIXME: This code is untestable due to __DIR__, also that hardcoded path is really dangerous
@@ -770,5 +770,18 @@
 		}
 		return null;
 	}
+    
+    /**
+     * Fixes major bug in Apache 2.4 / PHP 7 FPM where script_name contains full
+     * uri not just script name e.g /index.php/app/files instead of expected /index.php 
+     * @param string $name
+     * @return string
+     */
+    private function fixPhpFpmScriptName() {
+        if (strpos($this->server['SCRIPT_NAME'], '.php') !== false) {
+            $name = DIRECTORY_SEPARATOR . basename($this->server['SCRIPT_FILENAME']);
+        }
+        return $name;
+    }
 
 }
```

**Comment by DeepDiver1975** [comment]
I honestly cannot judge on this.
Since not many people are touched by this issue - can we move this to the FAQ section on central?

@RealRancor @tflidd objections?

**Comment by RealRancor** [comment]
@DeepDiver1975 A FAQ is always good as long as some one else is writing it. I won't do that as i don't have any idea about FPM on Apache.

And IMHO if some one wants to use FPM he should just use nginx which is known to work and has a good working config. :-)

**Comment by DeepDiver1975** [comment]
> A FAQ is always good as long as some one else is writing it. I won't do that as i don't have any idea about FPM on Apache.

hehe - same here :see_no_evil:

**Comment by Caffe1neAdd1ct** [comment]
@RealRancor @DeepDiver1975  I'm not just using Apache for owncloud and only want one webserver running on this setup due to limited hardware/resources. 

I think this is mainly due to using FPM through ProxyPassMatch from further research using SetHandler is the preferred method in Apache now which should also set the PATH_INFO SERVER variable correctly which would work with @matpen 's patch. 

I don't see anywhere in the docs stating only nginx can be used with owncloud so the accepted solution here seems contradictory to the official docs (https://doc.owncloud.org/server/9.1/admin_manual/installation/source_installation.html#apache-web-server-configuration)

It does look like the offical docs perfer php is used with apache through the mod_php implementation, however this locks your entire webserver to only one version of PHP which is a major limiting factor as not every application is ready for PHP 7 and still runs fine on PHP 5.6. FPM allows running every PHP version from 5.3 to 7 all on different ports :+1: 

So it would be good to look at fixing this for the SetHandler configuration at least as it does pass the PATH_INFO variable correctly.

**Comment by wpoely86** [comment]
@DeepDiver1975 That's not a real solution. It's the same as saying we should switch to nextcloud if we encounter a bug in owncloud...

I agree with @Caffe1neAdd1ct There are serious downsides to using mod_php with apache (like not being able to use the event based model). The fix is already upstream, the problem remains that it is not backported to RHEL7 & derivatives. A pragmatic solution where owncloud works around this until it is backported would be nice (instead of manually patching after every update).

**Comment by RealRancor** [comment]
@Caffe1neAdd1ct @wpoely86 I think everyone is happily accepting Pull Requests and a guarantee that you're maintaining the code path and that it doesn't break anything for existing users.

AFAIK officially only mod_php is supported and tested, everything else needs community maintainer.

**Comment by Caffe1neAdd1ct** [comment]
@wpoely86 Thanks for the update re backporting, i've been looking through the php-git / rhel bugzilla etc to find how close the backport is.. 

@RealRancor I'll have a look at a nicer patch solution and test with a few different server setups, i'm no alien to making pull requests :+1:

**Comment by wpoely86** [comment]
@Caffe1neAdd1ct I've already created a PR for this in the paste: #17491 Feel free to have a look, I haven't updated it recently. I've happy to test any patch you have. I currently run 9.0.4 on CentOS7 (with patch).

**Comment by Caffe1neAdd1ct** [comment]
@wpoely86 Cheers i'll take a look and test with my env.

My patch above is running fine with ProxyPassMatch set to `ProxyPassMatch ^/(.*\.php(/.*)?)$ fcgi://127.0.0.1:9001/usr/share/webapps/owncloud/$1
` and OwnCloud installed from Arch Linux Pacman version `9.1.0-1`

I want to test a good solution for working with both SetHandler and ProxyPassMatch to make sure most if not all Apache setups..

**Comment by Caffe1neAdd1ct** [comment]
@wpoely86 Your patch relies on `PATH_INFO` which isn't available with ProxyPassMatch, I found relying on `SCRIPT_FILENAME` worked across both setups...

**Comment by DeepDiver1975** [comment]
> @DeepDiver1975 That's not a real solution. 

Well - I simply cannot judge of the impact of the patches and I have to limit the risk of breaking hundreds or thousands of installations.



> It's the same as saying we should switch to nextcloud if we encounter a bug in owncloud...

I will not comment on that

**Comment by Caffe1neAdd1ct** [comment]
I've raised a pull for fixing on both ProxyPassMatch and SetHandler methods of using PHP-FPM through Apache 2.4 (https://github.com/owncloud/core/pull/25747).

I'm setting up environments now for testing through both of these setups and will look at testing through mod_php and nginx as well.

**Comment by kdslkdsaldsal** [comment]
FWIW Apache 2.4.26 broke php-fpm once again:

https://mail-archives.apache.org/mod_mbox/httpd-users/201706.mbox/%3c18166d8c-d58b-6c25-e90d-0a9f1f23cb35@gmail.com%3e

**Comment by PVince81** [comment]
was it fixed since ?

**Comment by kdslkdsaldsal** [comment]
Was fixed in 2.4.27 according to https://bz.apache.org/bugzilla/show_bug.cgi?id=61202

Lucky enough 2.4.26 wasn't available in Debian and Ubuntu, not sure about other Distros.
